frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,332 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactElement, useCallback, useEffect, useState } from "react"
+
+import { withTheme } from "@emotion/react"
+import WaveSurfer from "wavesurfer.js"
+import RecordPlugin from "wavesurfer.js/dist/plugins/record"
+import { Delete } from "@emotion-icons/material-outlined"
+
+import { FileUploadClient } from "@streamlit/lib/src/FileUploadClient"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import { AudioInput as AudioInputProto } from "@streamlit/lib/src/proto"
+import Toolbar, {
+  ToolbarAction,
+} from "@streamlit/lib/src/components/shared/Toolbar"
+import { EmotionTheme } from "@streamlit/lib/src/theme"
+import {
+  isNullOrUndefined,
+  labelVisibilityProtoValueToEnum,
+  notNullOrUndefined,
+} from "@streamlit/lib/src/util/utils"
+import { WidgetLabel } from "@streamlit/lib/src/components/widgets/BaseWidget"
+
+import { uploadFiles } from "./uploadFiles"
+import {
+  StyledAudioInputContainerDiv,
+  StyledWaveformContainerDiv,
+  StyledWaveformInnerDiv,
+  StyledWaveformTimeCode,
+  StyledWaveSurferDiv,
+} from "./styled-components"
+import NoMicPermissions from "./NoMicPermissions"
+import Placeholder from "./Placeholder"
+import {
+  BAR_GAP,
+  BAR_RADIUS,
+  BAR_WIDTH,
+  CURSOR_WIDTH,
+  STARTING_TIME_STRING,
+  WAVEFORM_HEIGHT,
+  WAVEFORM_PADDING,
+} from "./constants"
+import formatTime from "./formatTime"
+import ActionButton from "./ActionButton"
+
+interface Props {
+  element: AudioInputProto
+  uploadClient: FileUploadClient
+  widgetMgr: WidgetStateManager
+  theme: EmotionTheme
+}
+
+const AudioInput: React.FC<Props> = ({
+  element,
+  uploadClient,
+  widgetMgr,
+  theme,
+}): ReactElement => {
+  const [wavesurfer, setWavesurfer] = useState<WaveSurfer | null>(null)
+  const waveSurferRef = React.useRef<HTMLDivElement | null>(null)
+  const [deleteFileUrl, setDeleteFileUrl] = useState<string | null>(null)
+  const [recordPlugin, setRecordPlugin] = useState<RecordPlugin | null>(null)
+  // to eventually show the user the available audio devices
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [availableAudioDevices, setAvailableAudioDevices] = useState<
```

Comment:
is it possible to name the variable `_` instead of disabling `eslint` here as its possible in other languages? if yes, it might be a nicer convention

Reply:
tried that, but then our eslint naming convention rule complained about it not following naming conventions 😔

Reply:
to add on to this, i suspect that we will sooner rather than later need to add the ability to have some sort of drop down to select a mic. (and i'd rather not have to re-lookup how to do this 😅 )